Update python interface with rev2 features

achaikou opened the pull request:
1.
Python 3.9 removed in this PR because [original run](https://github.com/achaikou/segyio/actions/runs/19599389300) failed for some reason, no clue why.

2.
benchmarks are failing on my test runs.
header retrieval is slower for some reason, but most importantly "create" is way slower.
Looks to be caused by "Update Field to support trace headers" commit.

I don't see anything stupid, but please let me know if you do. We need to dig deeper as it is probably not a random failure.

Upd:
I managed for now to get slowdown to a more reasonable percentage, but the problem still stays.
Apparently `__setattr__` (and maybe `__getattr__` , didn't check that or the other classes with `.name` support) is just slow.
Simply defining it adds 25% slowness to 'create' benchmark test.

We need to discuss and find out if this is a deal breaker for our interface as Field is a main class responsible for fields extraction.

achaikou committed:
Add trace field names functionality

achaikou committed:
Add stanza mode to python interface

achaikou committed:
Fix sphinx config errors and warnings

achaikou committed:
Update Field to support trace headers

Field now supports all the traceheaders types, not just a standard one.

achaikou committed:
Update some outdated documentation

achaikou committed:
Drop support for python 3.9

Python 3.9 is EOL.

ajaust commented on python/test/segy.py:
```diff
@@ -980,6 +980,105 @@ def test_write_header(small):
+            f.traceheader[0][0][700]
+
+        with pytest.raises(KeyError):
+            f.traceheader[0][0][700] = 1
```
Does this test anything new compared to the test in line 991f?

ajaust commented on python/test/segy.py:
```diff
@@ -980,6 +980,105 @@ def test_write_header(small):
+        assert f.traceheader[1].SEG00000 == f.traceheader[0].SEG00000
+
+        f.traceheader[2][0:1] = f.traceheader[2][0:1]
+        assert list(f.traceheader[2][0:1]) == list(f.traceheader[2][0:1])
```
What are we testing here?

ajaust commented on python/segyio/field.py:
```diff
@@ -431,13 +474,51 @@ def __setitem__(self, key, val):
+        Examples
+        --------
+        Set a value:
+        >>> header.linetrc = 1006
```
I don't seem to find this docstring in my Sphinx docs either.

ajaust reviewed (commented):
This looks very nice overall. 🥳  I left some suggestions or questions here and there. Mostly about the docstrings. I also build the Sphinx documentation locally. The generated documenation looks like expected.


### Regarding the failing test in Python 3.9

It complains about `  FAILED test/segy.py::test_field_via_traceheader - Failed: DID NOT RAISE <clas..`. I am not sure if we should be worried that some expected exception is not thrown, but I am also not sure if investigating it further makes sense.

### Building with Sphinx

I build it with Python 3.14 and ran into the following issues

- `requirements-doc.txt` does not specify the ReadTheDocs theme as dependency. We would need to add `sphinx-rtd-theme` in the dependency file or one needs to use `requirements-readthedocs.txt` where the dependency is specified. We could also consider consolidating both the requirement files into one.
- Building the docs on failed since `conf.py` uses distutils for `LooseVersion`. `distutils` was deprecated in Python 3.10 and [removed in 3.12](https://docs.python.org/3/library/distutils.html). I replaced it with [`packaging.version.Version`](https://github.com/Azure/azure-cli/issues/17170) which resolved the issue.

achaikou commented on python/test/segy.py:
```diff
@@ -367,10 +367,10 @@ def test_disabled_methods_seismic_unix(openfn, kwargs):
+        assert 92 == len(list(f.header[0].keys()))
+        assert 92 == len(list(f.header[1].values()))
+        assert 92 == len(list(f.header[2].items()))
+        assert 92 == len(list(f.header[3]))
```
Yes.
Even those with separate of old Field and new Field I will revert those.
Let old code stay old.

achaikou commented on python/segyio/segy.py:
```diff
@@ -462,6 +463,55 @@ def header(self, val):
+        >>> traceheader.iline
+        ... 1005
+        >>> xline = traceheader.xline
+        ... 5
```
I probably first wrote it with SEG00000.

achaikou reviewed (commented):
Thanks for the blazing fast review! :fire: 

Python 3.9: I don't see a point in investigating it unless someone thinks it is useful.

Sphinx:
- last commit that touched `requirements-doc.txt` introduced it on purpose. I am not too deep into it to say if it was warranted or not, so I am inclined to leave it as is.
- thanks for the investigation! Applied.

----
### Performance

So running benchmark create test with `f.traceheader[tr][0].update({d})` is roughly 50% slower than `f.header[tr].update({d})`.

Reading this

    with segyio.open(filepath) as f:
        for tr in f.ilines * f.xlines:
            f.traceheader[tr][0][189]

vs
            
    with segyio.open(filepath) as f:
        for tr in f.ilines * f.xlines:
            f.header[tr][189]
            
is around 2 times slower.

From what I see slowness happens simply from default definition of `__setattr__`  (probably `__getattr__` too) for `TraceHeaderSequence` and `TraceHeaderField` interfaces, though I can't find any similar complaints online. I am running with 3.13 and all the issues seem to relate to 3.11
Maybe those are just fluctuations? Doubt that, given that it affected `header` test that is index-based, but still...

I am still not convinced we make the right decision to leave it be, because if slowness is real the only reasonable way to improve the performance is to remove the functions altogether, which would be a major breaking change for the users.

It probably must hold though, given that we don't have much time, but we need to highlight that in pre-release notes.

ajaust commented on python/segyio/field.py:
```diff
@@ -340,7 +340,7 @@ def __getitem__(self, key):
         Returns
         -------
-        value : int or dict_like
+        value : number, bytearray or dict_like
```
I also think I have see `number` and `number-like` so from my side you are free to decide what the best type is to describe the return type.

ajaust reviewed (approved):
Great! Thanks for the fixes. It looks good to me. I just left one or two comments as "unresolved" such that my reply is easier to find. I consider everything as resolved. I also really like that you updated the Sphinx build process. In this sense: "Approved" 🎉 

I understand that the performance issue is a potential problem. Do you have some pointer to the issues that were reported in Python 3.11? Should the performance issue also show up in our benchmark tests? At least the most recent benchmark run passed successfully.

yngve793 commented on python/segyio/field.py:
```diff
@@ -183,12 +183,10 @@ class Field(MutableMapping):
-        # segyfile and traceheader_index must be initialized before everything
-        # else because setattr is overridden
-        self.segyfile = segyfile
+    def __init__(self, segyfile, buf, kind, keys, traceno=None, traceheader_index=None):
```
It is probably well known: 
commit 1facd77d1ab855575d870a32d1b37992c554593b
Breaks test 
	  8 - python.unit (Failed)

achaikou replied:
As it is a fixup, no I had no clue.
But it is related to fixup-ing, should be fine with rebase.

yngve793 commented on python/segyio/field.py:
```diff
@@ -520,7 +601,7 @@ def update(self, *args, **kwargs):
         buf = bytearray(self.buf)
 
-        traceheader_index = 0
+        traceheader_index = self.traceheader_index
```
Can we just use` self.traceheader_index`?

yngve793 commented on python/segyio/trace.py:
```diff
@@ -773,6 +773,252 @@ def xline(self, value):
+            self[trace_index] = value
+
+
+class TraceHeaderSequence(Sequence):
```
Following the existing naming schemes it should not be necessary to specify that it is a Sequence.

What about `TraceHeaderDefintions` or `TraceHeaderSpesifications` ?

yngve793 commented on python/segyio/trace.py:
```diff
@@ -773,6 +773,252 @@ def xline(self, value):
+        """
+        if not isinstance(i, slice):
+            traceheader_index = self.wrapindex(i)
+            field_sequence = self[traceheader_index]
```
`field_sequence = self[self.wrapindex(i)]`?

achaikou replied:
Was ok in other place, but here I protest.
separate variable here gives me the context - I know what `self.wrapindex(i)` is supposed to mean now.

yngve793 commented on python/segyio/trace.py:
```diff
@@ -773,6 +773,252 @@ def xline(self, value):
             self.xline[i] = src
 
+
+class TracesTraceHeaderSequence(Sequence):
```
Not the best name. I have no real good alternative.  :-(
It seams like we are missing a term for the trace headers defined for a specific trace. 
If we had a name or word for this I think it would be easier to name this class.

yngve793 reviewed (approved):
In general the PR is fine. I have no strong objections. Some minor comments and suggestions are added. The only point I want to stress is the naming of classes in the  python/segyio/trace.py file. The current naming schemes works reasonably fine, but provides little conceptual understanding to the reader. What about something along these lines:

- TraceFieldEntry : Name, offset, type and value for a field
- TraceFieldSequence, TraceFieldList, TraceFieldDefinition, etc: Trace header definition as Standard, Extended etc.
- TraceFieldLayout : A sequence of TraceFieldDefinition containing Standard, Extended and private for a specific trace. 
- TraceFieldOverview: Sequence of TraceFieldLayouts for all traces. 

At least consider something along this lines.

achaikou committed:
Add Field wrapper to support access by names

New class had to be added to keep performance of the old Field class the
same.

Introduction of __getattr__ and __setattr__ directly to Field causes
decrease in performance compared with old code.

Nevertheless, we still want to give user the chance to retrieve data via
names and out of all approaches discussed in the team __getattr__ /
__setattr__ appeared to be the most pythonic and consistent.

So new class HeaderFieldAccessor was introduced on top of the Field one.

New naming scheme is chosen to
- support level-like access to the data
- distinguish this trace-number based hierarchy of field values from
trace-independent hierarchy of layout information
- alleviate confusion of names like Header/Field which actually are
lists/sequences/tables of headers/fields.

achaikou committed:
Add interface for access to all traceheaders

achaikou committed:
Support attributes operations on tracefield

achaikou reviewed (commented):
Thanks! :heart_eyes_cat: 

Name changes and other stuff is done directly and not in fixups, as we agreed.
Hopefully I updated everything correctly :smile:

ajaust commented on python/segyio/segy.py:
```diff
@@ -428,12 +428,12 @@ def header(self):
-        the iteration specified by the right-hand side (i.e. can skip headers
-        etc).
+        A convenient way for operating on all standard headers of a file is to
+        use the default full-file range.  It will write headers of trace 0, 1,
```
Not that important. It probably renders correctly, but there is a double space:
```suggestion
        use the default full-file range. It will write headers of trace 0, 1,
```

ajaust commented on python/segyio/field.py:
```diff
@@ -28,6 +28,9 @@ class Field(MutableMapping):
         more common dict operations (MutableMapping)
+
+    .. versionchanged:: 2.0
+       Support for SEG-Y revision 2.1
```
Just a temporary marker to remind me where I interrupted the review.

ajaust reviewed (approved):
Looks good to me. I only found 2 minor things. :grinning: 

I did not check the documentation website generate by sphinx after the rebase/squash yet. Just give me a sign if you want me to do that. :mag:

ajaust reviewed (approved):
I just verified that the docstrings appear in the Sphinx documentation. I found some minor things (I am sorry) and added them to the review.

achaikou commented on python/segyio/trace.py:
```diff
@@ -992,3 +992,85 @@ def __setitem__(self, i, val):
+        """
+        return self._names
+
+    def __getitem__(self, key):
```
I think we should leave it as is - after all old classes have the same ordering.

Probably if we want to do something about it, we should do that for the whole documentation.
But I don't see any problem or need in that.

Let me know if you feel different!

achaikou commented on python/segyio/trace.py:
```diff
@@ -1596,3 +1598,27 @@ def __str__(self):
+
+        Returns
+        -------
+        array of numbers//bytearray
```
No, this is me hating this amount of documentation and making so many typos that I don't see them anymore :smile:

achaikou reviewed (commented):
Thanks for finding new typos! :tada:

yngve793 reviewed (approved):
Fixups looks fine to me. 
No further comments.

ajaust reviewed (approved):
Looks good. 🎉 
Go ahead and apply the fixups and merge.
